vehicle_gateway_demo/src/vtol_position_control.cpp

```diff
@@ -0,0 +1,238 @@
+    double airspeeed = 15.0,
+    double distance_threshold = 10.0,
+    vehicle_gateway::CONTROLLER_TYPE controller_type = vehicle_gateway::CONTROLLER_TYPE::POSITION)
+  {
+    while (true) {
+      this->gateway_->set_offboard_control_mode(vehicle_gateway::CONTROLLER_TYPE::POSITION);
+      this->gateway_->set_local_position_setpoint(x, y, alt, yaw);
+      this->gateway_->set_airspeed(airspeeed);
+
+      float current_ned_x, current_ned_y, current_ned_z;
+      this->gateway_->get_local_position(current_ned_x, current_ned_y, current_ned_z);
+      const auto dx = x - current_ned_x;
+      const auto dy = y - current_ned_y;
+      const auto dalt = alt - current_ned_z;
+      const auto distance = sqrt(pow(dx, 2) + pow(dy, 2) + pow(dalt, 2));
+
+      if (distance < distance_threshold) {
+        break;
+      }
+
+      usleep(1e5); // 100 ms
+    }
+  }
+
+  void transition_to_offboard_sync()
+  {
+    for (int i = 0; i < 20; i++) {
+      this->gateway_->set_local_position_setpoint(
+        0.0, 0.0,
+        numeric_limits<float>::quiet_NaN(), numeric_limits<float>::quiet_NaN());
+      this->gateway_->set_offboard_control_mode(vehicle_gateway::CONTROLLER_TYPE::POSITION);
+      usleep(1e5);  // 100 ms
+    }
+    this->gateway_->set_offboard_mode();
+  }
+
+  void transition_to_multicopter_sync()
+  {
+    while (this->gateway_->get_vtol_state() != vehicle_gateway::VTOL_STATE::MC) {
+      this->gateway_->transition_to_mc();
+      usleep(1e5);  // 100 ms
+    }
+  }
+
+  void transition_to_fixed_wing_sync()
+  {
+    while (this->gateway_->get_vtol_state() != vehicle_gateway::VTOL_STATE::FW) {
+      this->gateway_->transition_to_fw();
+      usleep(1e5);  // 100 ms
+    }
+  }
+
+  void destroy()
+  {
+    if (this->gateway_) {
+      this->gateway_->destroy();
+      this->gateway_ = nullptr;
+    }
+    if (this->loader_) {
+      this->loader_->unloadLibraryForClass(
+        "vehicle_gateway_px4::VehicleGatewayPX4");
+      this->loader_ = nullptr;
+    }
+  }
+
+  std::shared_ptr<vehicle_gateway::VehicleGateway> gateway_;
+
+private:
+  std::shared_ptr<pluginlib::ClassLoader<vehicle_gateway::VehicleGateway>> loader_;
+};
+
+
+int main(int argc, const char * argv[])
+{
+  rclcpp::init(argc, argv);
+
+  const auto vg = std::make_shared<VehicleGatewayCpp>();
+  const float TARGET_ATTITUDE = 30.0;
+
+  cout << "Arming..." << endl;
+  vg->arm_sync();
+  sleep(2);
+
+  const auto home_position = vg->gateway_->get_latlon();
+  if (home_position.size() != 3) {
+    throw std::runtime_error("home_position should have three elements: lat, lon, alt");
+  }
+  cout << "Home altitude: " << home_position[2] << endl;
+
+  cout << "Takeoff!" << endl;
+  vg->gateway_->takeoff();
+  sleep(1);
+
+  vg->go_to_latlon_sync(home_position[0], home_position[1], home_position[2] + TARGET_ATTITUDE);
+
+  vg->transition_to_fixed_wing_sync();
+
+  cout << "Begin transitioning to Offboard control..." << endl;
+  vg->transition_to_offboard_sync();
```

Comment:
I made this a function and pass NaN as an arg to altitude so it's not necessary.